### PR TITLE
Dont precompile already seen keys, limit epilogue choices

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124031
* #122825
* #123229
* __->__ #122642
* #124030

Two changes: 
- in epilogue benchmark fusion, only take top 6 choices. There were basically no choices taken after this in HF.
- Share a single precompilation function among matmuls with same key. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Review by shunting314 — COMMENTED** [review]
Can we share a bit of the impact on compilation time and perf? If a dashboard run takes too long, some local run results for a few models (or microbenchmarks) should also be good.

File: torch/_inductor/select_algorithm.py [hunk 1]
```diff
@@ -844,6 +853,8 @@ def __call__(
         def make_benchmark_fn():
             return self.make_benchmark_fn(choices, input_nodes, layout, input_gen_fns)
 
+        inputs_key = repr([self.key_of(x) for x in input_nodes])
```
[the remark this answers is not in the document]
Comment (shunting314) [reply torch/_inductor/select_algorithm.py:856]: Give it another thoughts, mm/addmm/bmm happens to have different input metadatas... But would be nice to add op name to the precompile cache key.

**Review by shunting314 — APPROVED** [review]

File: torch/_inductor/select_algorithm.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (eellison) [reply torch/_inductor/select_algorithm.py:914]: added test, and name and precision in key

**Pull request closed without merging**